packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1215,32 +1270,161 @@ export function errorHydratingContainer(parentContainer: Container): void {
+  instance: Instance,
   internalInstanceHandle: Object,
 ): void {
-  if (enableScopeAPI) {
-    precacheFiberNode(internalInstanceHandle, scopeInstance);
+  switch (type) {
+    case 'html':
+    case 'head':
+    case 'body': {
+      clearSingletonInstance(instance);
+      resetProperties(instance, type, props);
+      precacheFiberNode(internalInstanceHandle, instance);
+      updateFiberProps(instance, props);
+      return;
+    }
+    default: {
+      throw new Error(
+        'commitSingletonPlacement was called with an element type that is not supported. This is a bug in React.',
+      );
+    }
   }
 }
 
-export function getInstanceFromScope(
-  scopeInstance: ReactScopeInstance,
-): null | Object {
-  if (enableScopeAPI) {
-    return getFiberFromScopeInstance(scopeInstance);
+export function getInsertionEdge(parent: Instance): ?InstanceSibling {
+  if (enableFloat) {
+    let node = null;
+    let nextNode = parent.lastChild;
+    let fallbackNode;
+    while (nextNode != null) {
+      const fiber = getInstanceFromNodeDOMTree(nextNode);
+      if (fiber) {
+        // We intentionally start with fiber rather than fiber.return because we want to
+        // account whether the node we found is the root itself. This comes into play
+        // when you portal into the same element that contains the HostRoot
+        let parentFiber = fiber;
+        while (parentFiber !== null) {
+          if (
+            (parentFiber.tag === HostComponent &&
+              parentFiber.stateNode === parent) ||
+            ((parentFiber.tag === HostPortal || parentFiber.tag === HostRoot) &&
+              parentFiber.stateNode.containerInfo === parent)
+          ) {
+            return node;
+          }
+          if (fallbackNode === undefined && parentFiber.tag === HostRoot) {
+            // When we find out first Fiber Node we capture the preceding Node to use as a fallback
+            // This is because we want to append to sibling fiber trees but prepend non-fiber trees
+            // If we don't end up finding an explicit insertion point based on existing siblings
+            fallbackNode = fallbackNode || node;
+          }
+          parentFiber = parentFiber.return;
+        }
+      }
+
+      node = nextNode;
+      nextNode = nextNode.previousSibling;
+    }
+    // We return the fallbackNode if we found one (the Node following the first React owned Node)
+    // Otherwise we return the first Node in the list of Siblings
+    return fallbackNode !== undefined ? fallbackNode : node;
   }
   return null;
 }
 
+function clearSingletonInstance(instance: Instance) {
+  const tagName = instance.tagName.toLowerCase();
+  switch (tagName) {
+    case 'html': {
+      return;
+    }
+    case 'head':
+    case 'body': {
+      let node = instance.firstChild;
+      while (node) {
+        const nextNode = node.nextSibling;
+        const nodeName = node.nodeName.toLowerCase();
+        if (
+          nodeName === STYLE ||
```

Comment:
This STYLE is not the node name but the property name `.style` so don't use this constant. It's probably safe to assume that this is a HTML document and that they're going to be upper case too.